02-intermediate-ruby/intro-to-minitest.md

```diff
@@ -0,0 +1,378 @@
+
+```ruby
+# bill.rb
+
+class Bill
+
+end
+```
+
+Now `ruby bill_spec.rb` gets us:
+
+```bash
+Run options: --seed 32657
+
+# Running:
+
+Run options: --seed 32657
+
+# Running:
+
+..
+
+Finished in 0.001089s, 918.2736 runs/s, 918.2736 assertions/s.
+
+1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
+
+Finished in 0.454822s, 2.1987 runs/s, 2.1987 assertions/s.
+1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
+
+```
+
+Now we have our first **green/passing** test.
+
+There are a [number of expectations](http://mattsears.com/articles/2011/12/10/minitest-quick-reference/) in Minitest beyond the `must_equal` method.
+
+**Exercise** Look at the list of expectations above and revise the `bill.class.must_equal Bill` to use the `must_be_instance_of` expectation instead.
+
+
+#### A Word on Parentheses
+
+In the code above, we are calling `must_equal` without using parentheses `bill.class.must_equal Bill`.  Ruby doesn't FORCE you to put parentheses around a method's arguments but [the community-driven style guidelines](https://github.com/bbatsov/ruby-style-guide) suggest that it's good coding style to put parentheses around method arguments **except** for methods which are part of an internal Domain Specific Language (DSL), or basically the syntax of some kind of framework like... Minitest.
+
+So you shouldn't put parentheses around the arguments to `must` method arguments, but you **should** around your own methods.
+
+### Adding Functionality
+
+We are going to require that Bill must be instantiated with a list (array) of hashes.  Each item will have a name and price.  So we will revise the test to look like this:
+
+```ruby
+# bill_test.rb
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'bill'
+
+describe "bill" do
+  it "can be created" do
+    bill = Bill.new [{name: 'Coffee', price: 4.75}, {name: 'Pancakes', price: 8.75}]
+    # the class of @bill should be Die
+    bill.class.must_equal Bill
+  end
+end
+```
+
+We also need to ensure that if Bill.new is called without an argument it raises an error.  
+
+So we will add another `it` block and use the `must_raise` expectation.
+
+```ruby
+it "will raise an error if created with an invalid argument" do
+  proc {
+    bill = Bill.new
+  }.must_raise ArgumentError
+end
+```
+
+**Question**: Look back at the [guidelines for what to test]().  What other kinds of things should we test on the `initialize` method?
+
+#### Practice Exercise
+
+Now we need make a method that calculates the bill subtotal.  Write a test (`it` block) that creates an instance of Bill, calls a `subtotal` method and ensure that it returns the correct answer (the sum of the prices).  You can use the `must_equal` matcher.  
+
+#### Check & Verify
+
+Check with your neighbor.  You can find a solution [here.](https://github.com/AdaGold/bill_calculator/commit/12548edbe1d88bce598b642764e7953c10ced204)
+
+### Arrange-Act-Assert
+
+In our tests we will typically follow this pattern:
```

Comment:
Would it make sense to move this section earlier in the resource? It seems a little odd to write a few tests and then talk about what order to do things in a test.